### PR TITLE
Vision a11 - adding crop hints and web annotation

### PR DESCRIPTION
**Pull request by puneith**
@jonparrott @tswast Please review. Alpha 1.1 changes.

### CODE REVIEW

File: vision/api/label/snippets.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+# Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (tswast) [inline vision/api/label/snippets.py]: 2016.

Also, I don't think the header on go/releasing/preparing includes "All Rights Reserved" anymore.

File: vision/api/label/snippets.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    scoped_credentials = credentials.create_scoped(
+        ['https://www.googleapis.com/auth/cloud-platform'])
+    http = httplib2.Http()
+    scoped_credentials.authorize(http)
+    return discovery.build('vision', 'v1',
+                           http=http,
+                           discoveryServiceUrl=DISCOVERY_URL)
+
+
+def crop_hint(photo_file):
```
Comment (tswast) [inline vision/api/label/snippets.py]: `photo_file` makes me think this is a file object. Prefer `photo_path`.

(Update here and in the other functions)

File: vision/api/label/snippets.py (hunk 1 above)
Comment (puneith) [inline vision/api/label/snippets.py]: done

File: vision/api/label/snippets.py (hunk 2 above)
Comment (puneith) [inline vision/api/label/snippets.py]: done

**Comment by puneith** [comment]
@jonparrott The tests are successful locally. Any ideas why one of the three fails?

**Comment by jonparrott** [comment]
@puneith the response list is empty:

```
(Pdb) result
{u'responses': [{}]}
```

File: vision/api/label/snippets.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import argparse
+import base64
+import httplib2
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: httplib2 is a third-party import.

File: vision/api/label/snippets.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import json
+
+from googleapiclient import discovery
+from oauth2client.client import GoogleCredentials
+
+DISCOVERY_URL = ('https://vision.googleapis.com/$discovery/rest?'
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: Break after opening `(`, please.

File: vision/api/label/snippets.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                 'labels=TRUSTED_TESTER&version=v1')
+
+
+def get_service():
+    """Get vision service using discovery."""
+    credentials = GoogleCredentials.get_application_default()
+    scoped_credentials = credentials.create_scoped(
+        ['https://www.googleapis.com/auth/cloud-platform'])
+    http = httplib2.Http()
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: I know I ask this constantly, but is this needed because of alpha?

File: vision/api/label/snippets.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    scoped_credentials.authorize(http)
+    return discovery.build('vision', 'v1',
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: Break after opening `(`, please.

File: vision/api/label/snippets.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def crop_hint(photo_path):
+    """Run a crop hint request on the image."""
+
+    service = get_service()
+
+    with open(photo_path, 'rb') as image:
+        image_content = base64.b64encode(image.read())
+        service_request = service.images().annotate(body={
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: Your `with` statement covers too many things. It only needs to cover the `image.read` call.

File: vision/api/label/snippets.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        return response
+
+
+def web_annotation(photo_path):
+    """Run a web annotation request on the image."""
+
+    service = get_service()
+
+    with open(photo_path, 'rb') as image:
+        image_content = base64.b64encode(image.read())
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: Same comment here on the with statement.

File: vision/api/label/snippets.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('command', choices=['crop_hint', 'web_annotation'])
+    parser.add_argument('image_path', help='The image you\'d like to process.')
+    args = parser.parse_args()
+
+    if args.command == 'crop_hint':
+        response = crop_hint(args.image_path)
+        print(json.dumps(response, indent=2))
```
Comment (jonparrott) [inline vision/api/label/snippets.py]: Prefer printing the result in the sample functions themselves.

File: vision/api/label/snippets_test.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    assert len(result['responses']) == 1
+
+
+def test_crop_hint_response_dim(resource):
+    result = snippets.crop_hint(resource('cat.jpg'))
+    crop_hint = result['responses'][0]
+    crop_hint_annotation = crop_hint['cropHintsAnnotation']['cropHints'][0]
+    confidence = crop_hint_annotation['confidence']
+
+    assert confidence > 0.5
```
Comment (jonparrott) [inline vision/api/label/snippets_test.py]: Try `assert 0.5 < confidence < 0.9`

**Comment by puneith** [comment]
@gguuss fyi

File: vision/api/label/snippets.py (hunk 5 above)
Comment (puneith) [inline vision/api/label/snippets.py]: yes

File: vision/api/label/snippets.py (hunk 9 above)
Comment (jonparrott) [inline vision/api/label/snippets.py]: Our samples should not be structured around our tests, it should be the other way around. Please be consistent with the other samples.

**Comment by jonparrott** [comment]
@tswast can you approve changes?

**Comment by puneith** [comment]
@jonparrott Now that this is finally green can you please merge.

**Comment by puneith** [comment]
@dpebot Do I have powers to ask you to merge?

**Pull request merged**